### PR TITLE
Port over relevant tests and code from parity-hash

### PR DESCRIPTION
**Pull request by dvdplm**
The `pwasm-*` crates currently use [`parity-hash`](https://github.com/fckt/parity-hash) while `parity-ethereum` and `substrate` use `fixed-hash`. This PR tries to rectify this by porting over relevant changes from `parity-hash` to `fixed-hash` with the end goal of deprecating the former. One important difference is that `fixed-hash` does not export any concrete types so in order to stop using `parity-hash` the type construction should be added to one of the `pwasm-*` crates (`pwasm-abi` perhaps?).

### CODE REVIEW

File: fixed-hash/src/hash.rs [hunk 1]
```diff
@@ -321,6 +335,80 @@ macro_rules! construct_hash {
...
+				let mut ret = $a::new();
+				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(value);
+				ret
+			}
+		}
+	}
+}
+
+/// Implements conversion to and from a hash type and the equally sized unsigned int.
+/// CAUTION: Bad things will happen if the two types are not of the same size!
```
Comment (NikVolf) [inline fixed-hash/src/hash.rs]: maybe add a debug_assert along with caution?

Comment (dvdplm) [reply fixed-hash/src/hash.rs]: Sort of like I did [here](https://github.com/paritytech/rust-rocksdb)? What are your thoughts on `assert` vs `debug_assert` in this case?

Comment (NikVolf) [reply fixed-hash/src/hash.rs]: regular `assert` is not zero cost in production i believe

Comment (dvdplm) [reply fixed-hash/src/hash.rs]: Right, I was more asking if it's worth it to use `assert!` anyway or if the cost is prohibitive.

Comment (pepyakin) [reply fixed-hash/src/hash.rs]: what bad things ought to happen exactly? Isn’t this bad things already encompass panicking?

Comment (dvdplm) [reply fixed-hash/src/hash.rs]: @pepyakin you are correct! :) Added a `debug_assert`, which is similar to what `uint` does.

**Pull request merged**